### DISCUSSION TITLE
Refactoring CICD for GitLab and licensing question

### DISCUSSION THREAD

**Discussion by timothystone** [jhipster/generator-jhipster#27829] [Q&A]
I'm undertaking a refactoring of the GitLab templates in the `ci-cd` subgenerator. As part of this refactoring, I'm considering using the excellent work of the To Be Continuous (TBC) project, TBC is "a set of GitLab CI templates developed and maintained by DevOps and technology experts to build state-of-the-art CI/CD pipelines in minutes." The TBC templates were recently made available to the GitLab CI/CD Catalog.

I'm a contributor to TBC, having written the Maven Jib template variant—to support our enterprise use of JHipster!—and besides seeing some overdue updates needed in the `ci-cd` subgenerator as a whole, i.e., should Artifactory really be the default in the sub generator, when all that is being set in the `distributionManagement:url`? If it's a GitLab choice, default to GitLab registries—which is [the TBC default behavior](https://to-be-continuous.gitlab.io/doc/dev/architecture/#gitlab-by-default) by design—and provide documentation for using Artifactory, Nexus, Archiva (Apache), &etc.

The question I have is how JHipster is licensed vs. TBC. TBC is licensed with the LGPL. Before we pick up our torches and wander into the rabbit hole of "viral-ness" I want to see if there is a general consensus that using TBC as licensed _is not a show-stopper_.

I have had informal (non-retainer) conversations with an IP-in-IT lawyer friend (we're actually quite close as he is my gamemaster in a multi-year RPG campaign). He advises that following the LGPL §4 Combined Works I should include the LGPL license file and note the use of LGPL code in the generator file and that the refactored code would have a boundary on the "viral-ness" to the GitLab template of the generator. TBC is agnostic across the board on all things if only taking a GitLab-first approach.

**Comment by TatyOko28** [comment]
This sounds like a valuable modernization effort for the `ci-cd` subgenerator, especially if the goal is to align JHipster’s GitLab templates with current GitLab-native CI/CD practices.

## About Licensing (JHipster vs TBC LGPL)

JHipster is primarily distributed under the **Apache License 2.0**.

The **To Be Continuous (TBC)** templates being licensed under **LGPL** does not automatically make them unusable in JHipster, but it does mean the integration should be evaluated carefully.

Your lawyer friend is pointing to the correct area: **LGPL combined works / aggregation / boundary separation** is usually the key issue.

If JHipster is:

```text
- generating templates that originate from TBC
- referencing external TBC templates
- adapting TBC snippets while preserving attribution
- keeping a clear separation between generator code and template content

**Reply by timothystone** [reply]
Thanks @TatyOko28 for touching this topic. Some work I left undone and should revisit.

JHipster would generate a `.gitlab-ci.yml` file that used TBC components and set values to the used TBC components, such as follows:

```
include:
  # 1: include the component
  - component: $CI_SERVER_FQDN/to-be-continuous/maven/gitlab-ci-maven@5.4.0
    # 2: set/override component inputs
    inputs:
      # ⚠ this is only an example
      image: docker.io/library/maven:3.8-openjdk-18
      deploy-enabled: true
```

As to each of the noted items:

- generating templates that originate from TBC

No new templates for TBC would be generated. Only the expected GitLab CI descriptor that leveraged TBC.

- referencing external TBC templates

As shown above, yes.

- adapting TBC snippets while preserving attribution

Attribution would be included with the LGPL license and can be noted in the generated `.gitlab-ci.yml` file as well.

- keeping a clear separation between generator code and template content

That would be clear as the generated `.gitlab-ci.yml` would be the only artifact.
